libvast/vast/fbs/utils.hpp

```diff
@@ -71,11 +79,57 @@ pack_bytes(flatbuffers::FlatBufferBuilder& builder, const T& x) {
   return builder.CreateVector(data, bytes.size());
 }
 
-caf::expected<flatbuffers::Offset<TableSliceBuffer>>
-pack(flatbuffers::FlatBufferBuilder& builder, table_slice_ptr x);
+/// Generic unpacking utility.
+/// @tparam Flatbuffer The flatbuffer type to unpack.
+/// @param xs The buffer to unpack a flatbuffer from.
+/// @returns A pointer to the unpacked flatbuffer of type `Flatbuffer` or
+///          `nullptr` if verification failed.
+/// @relates unverified_unpack
+template <class Flatbuffer, size_t Extent = dynamic_extent>
+const Flatbuffer* as_flatbuffer(span<const byte, Extent> xs) {
+  /// The flatbuffer file_identifier. Defining this identifier here is
+  /// workaround for the lack of traits for generated flatbuffer classes. The
+  /// only way to get a flatbuffer identifier is by calling MyTypeIdentifier().
+  /// Because this prevents use of generic programming, we only use one global
+  /// file identifier for VAST flatbuffers.
+  constexpr const char* file_identifier = "VAST";
+  // Verify the buffer.
+  auto data = reinterpret_cast<const uint8_t*>(xs.data());
+  auto size = xs.size();
+  char const* identifier = nullptr;
+  if (flatbuffers::BufferHasIdentifier(data, file_identifier))
+    identifier = file_identifier;
+  flatbuffers::Verifier verifier{data, size};
+  if (!verifier.template VerifyBuffer<Flatbuffer>(identifier))
+    return nullptr;
+  return flatbuffers::GetRoot<Flatbuffer>(data);
+}
+
+/// Wraps an object into a flatbuffer.
+template <class T>
+caf::expected<chunk_ptr>
+wrap(T const& x, const char* file_identifier = nullptr) {
```

Comment:
The `file_identifier` thing is not clear to me. Why can it be `nullptr`? If we want to compose builders for nested tables it should happen inside of `pack`, no?

Reply:
I've added a note in the documentation. Let me know whether that suffices.

Reply:
The comments clarify it somewhat, but I still have to question why we can't drop the `file_identifier` argument and hard code `"VAST"` when calling `Finish()`?

Reply:
Because not *every* flatbuffer has a file identifier. Only those that are *root* buffers, i.e., those that begin the file.

Given a flatbuffer, you can figure out whether it has a `file_identifier`, but not the other way round. It's a limitation of the API because it was not built to support type-conditional packing in a generic way.

Reply:
> Because not every flatbuffer has a file identifier. Only those that are root buffers, i.e., those that **begin** the file. (emphasis mine)

So we can write multiple independent flatbuffers into one file instead of defining a `table` that contains all of those and declaring that the root object?

Or to look at this from a different angle: What is the use case for calling `wrap` on an object that is not going to be the top-level item in a file? Currently there is no such use in the code.

Reply:
> So we can write multiple independent flatbuffers into one file instead of defining a table that contains all of those and declaring that the root object?

The file identifier has nothing to do with the scenario when multiple flatbuffers are written to the same file. When specified in a *.fbs file, `flatc` will generate special functions for finishing the flatbuffer such that it starts with the identifier. If we used the these generated functions, like `FinishXXX`, then it would be completely transparent. However, `pack` and `unpack` are generic. They can't call a generated function. They can only dispatch on types. Because [there are no traits for this scenario](https://groups.google.com/d/msg/flatbuffers/DzSpIE-6fRA/Zm544dxvAgAJ), it's our burden to add the identifier manually, based on whether we have a buffer that's to be persisted on disk or whether a buffer that just gets composed into another flatbuffer.

> What is the use case for calling wrap on an object that is not going to be the top-level item in a file? Currently there is no such use in the code.

This will be a table slice inside segments, for example.

Reply:
Here is a snippet [from the docs](https://google.github.io/flatbuffers/md__schemas.html):

> Typically, a FlatBuffer binary buffer is not self-describing, i.e. it needs you to know its schema to parse it correctly. But if you want to use a FlatBuffer as a file format, it would be convenient to be able to have a "magic number" in there, like most file formats have, to be able to do a sanity check to see if you're reading the kind of file you're expecting.
> 
> Now, you can always prefix a FlatBuffer with your own file header, but FlatBuffers has a built-in way to add an identifier to a FlatBuffer that takes up minimal space, and keeps the buffer compatible with buffers that don't have such an identifier.
> 
> You can specify in a schema, similar to `root_type`, that you intend for this type of FlatBuffer to be used as a file format:

Reply:
FWIW, I've [filed a PR](https://github.com/google/flatbuffers/pull/5904) in the flatbuffers repo to fix this.

Reply:
> > What is the use case for calling wrap on an object that is not going to be the top-level item in a file? Currently there is no such use in the code.

> This will be a table slice inside segments, for example.

I see now, we pre-construct `TableSlice`s at the source, but they don't need a file identifier, because the `Segment` provides that already.